python/lsst/afw/geom/calculateSipWcsHeader.py

```diff
@@ -0,0 +1,80 @@
+from lsst.geom import Extent2I, Box2D
+from lsst.daf.base import PropertyList
+
+__all__ = ("calculateSipWcsHeader",)
+
+
+def coefficientsToHeader(order, coeffMatrix, header, name):
+    """Convert SIP coefficients to header keywords
+
+    Parameters
+    ----------
+    order : `int`
+        SIP order (equal to the maximum sum of the polynomial exponents).
+    coeffMatrix : `numpy.ndarray`
+        A square matrix of polynomial coefficients. ``coeffMatrix[i][j]``
+        is the coefficient of ``x^i y^j``.
+    header : `lsst.daf.base.PropertyList`
+        FITS header to which to add SIP coefficient keywords.
+    name : `str`
+        Name of polynomial. For the SIP FITS convention, these are one of
+        ``A``, ``B``, ``AP`` and ``BP``.
+    """
+    header.set(f"{name}_ORDER", order)
+    for xx in range(order + 1):
+        for yy in range(order + 1 - xx):
+            header.set(f"{name}_{xx}_{yy}", coeffMatrix[xx][yy])
+
+
+def calculateSipWcsHeader(wcs, order, bbox, spacing, header=None):
+    """Generate a SIP WCS header approximating a given ``SkyWcs``
+
+    Parameters
+    ----------
+    wcs : `lsst.afw.geom.SkyWcs`
+        World Coordinate System to approximate as SIP.
+    order : `int`
+        SIP order (equal to the maximum sum of the polynomial exponents).
+    bbox : `lsst.geom.Box2I`
+        Bounding box over which to approximate the ``wcs``.
+    spacing : `float`
+        Spacing between sample points.
+    header : `lsst.daf.base.PropertyList`, optional
+        Header to which to add SIP WCS keywords.
+
+    Returns
+    -------
+    header : `lsst.daf.base.PropertyList`
+        Header including SIP WCS keywords.
+
+    Examples
+    --------
+    >>> header = calculateSipWcsHeader(exposure.getWcs(), 3, exposure.getBBox(), 20)
+    >>> sipWcs = SkyWcs(header)
+    """
+    transform = getPixelToIntermediateWorldCoords(wcs)
+    crpix = wcs.getPixelOrigin()
+    cdMatrix = wcs.getCdMatrix()
+    crval = wcs.getSkyOrigin()
+    gridNum = Extent2I(int(bbox.getWidth()/spacing + 0.5), int(bbox.getHeight()/spacing + 0.5))
+
+    sip = SipApproximation(transform, crpix, cdMatrix, Box2D(bbox), gridNum, order)
+
+    if header is None:
+        header = PropertyList()
+    header.set("CTYPE1", "RA---TAN-SIP")
+    header.set("CTYPE2", "DEC--TAN-SIP")
+    header.set("CRVAL1", crval.getRa().asDegrees())
+    header.set("CRVAL2", crval.getDec().asDegrees())
+    header.set("CRPIX1", sip.getPixelOrigin().getX() + 1)  # +1 for LSST --> FITS
+    header.set("CRPIX2", sip.getPixelOrigin().getY() + 1)  # +1 for LSST --> FITS
+    for xx, yy in itertools.product(range(2), range(2)):
+        header.set(f"CD{xx + 1}_{yy + 1}", sip.getCdMatrix()[xx][yy])
+    coefficientsToHeader(sip.getOrder(), sip.getA(), header, "A")
+    coefficientsToHeader(sip.getOrder(), sip.getB(), header, "B")
+    coefficientsToHeader(sip.getOrder(), sip.getAP(), header, "AP")
+    coefficientsToHeader(sip.getOrder(), sip.getBP(), header, "BP")
```

Comment:
instead of modifying the header directly, did you consider using `makeTanSipWcs` (defined in `afw/geom/SkyWcs.h`) to make a TAN-SIP SkyWcs, and then using `getFitsMetadata()` on that?  I could imagine that being a fair bit slower, as it'll initialize a bunch of AST stuff, but it might be better at handling FITS edge cases.  There's nothing obviously wrong with your code here, but I'm a bit worried about non-obvious things.

Reply:
Oh. I didn't realise that was available...
But it doesn't work out-of-the-box. I'm investigating...

Reply:
After another fix to `SipApproximation`, it's working.